### DISCUSSION TITLE
Notemap Viewer

### DISCUSSION THREAD

**Discussion by gerdb** [cardonabits/haxo-rs#18] [Show and tell]
I've written a viewer for the notemap.json file in JavaScript:
[Notemap Viewer](https://www.sebulli.com/haxo/notemap/)

![NotemapViewer](https://github.com/cardonabits/haxo-rs/assets/883695/4c09844b-fefe-4a25-a170-b99cf3b777a2)

**Comment by jcard0na** [comment]
Nice!
I wrote a similar take on the same:  https://haxo-notemap.nn.r.appspot.com/
I think I like yours better!

**Reply by gerdb** [reply]
:) But yours is more than just a viewer.

**Reply by jcard0na** [reply]
Yes, but yours looks better.  I'd be happy to give you access to that repo if there is any code you can reuse for your viewer.  Let me know.

**Reply by thuntjazz** [reply]
Hey guys - I think the B & bis Bb keys are labeled backwards on the appspot version?

Otherwise very nicely done - and thanks for helping me figure out the top F# fingering  :-)

TH

**Reply by jcard0na** [reply]
> I think the B & bis Bb keys are labeled backwards on the appspot version?

Yes, you are right.  I just fixed it.  Thanks!

**Reply by thuntjazz** [reply]
Nice!  I also like the change to buttons for the alternate fingerings, rather than a pull-down menu.

I might suggest one more thing...prioritizing the fingerings based on their common usage on a saxophone.  For example, the 3rd space C lists 3 possible fingerings: the chromatic "side C" is the default that pops up, followed by the same fingering also covering the bis Bb as alternate 1, then the "normal" middle finger C as alternate 2.  I would suggest making middle finger C alternate 0 (default).  Another example would be F#: middle finger as the default and the chromatic as a later alternate.

I'm no help at all with the coding and Linux stuff, but I hope I can contribute meaningful stuff from a performer/end user perspective.  :-)

**Reply by Ekolide** [reply]
@jcard0na I've found that when deleting all fingerings for a note, then adding a new fingering destroys the UI with no apparent way to get it back.

**Reply by jcard0na** [reply]
Thanks for reporting.

**Reply by Ekolide** [reply]
Hey I have some questions about the implementation of the notemap. I haven't done all my tests on the latest image so I might be out of date.

1. How picky is the Haxophone with the list ordering in the notemap.json? I tried inputting fingerings for low A and Ab as well as high G (maybe G# too), but it would not load. I did simply put the fingerings manually at the end of the list, would that be part of the problem or is the Haxophone somehow limited in the note range it accepts in the notemap?
2. Could I make a C-transposed notemap by simply offsetting the MIDI Notes in the notemap? Of course I would love to select it out of the dropdown menu on the web editor, but I might be able to script it for myself. With the coming changes to transposing, maybe this doesn't matter?
3. I'm working on some notemaps that are very large. Do you think these would affect the realtime performance of the Haxophone? One of the notemaps uses ca 25-30 fingerings for the same note. Another one as of now has just over 80 fingerings for the same note, though I don't know if it will go much above 90-100 fingerings per note.

Some reflections and ideas on the notemap editor.

1. It's specific about only loading and saving files called `notemap.json`. This makes it hard to work on multiple notemap projects. A "Save As" function, and being able to load files with arbitrary names like `EWI_01.json`, `FluteNotemap.json`, or `C-transposed notemap.json` would be beneficial.
2. Extended notes. As mentioned, I'm working on some fingerings that gets the Haxophone to a low A and Ab, as well as a high G and maybe G#. I would love to implement these through the web editor, so I don't have to manually go editing and sorting.

Thanks!

**Reply by jcard0na** [reply]
> How picky is the Haxophone with the list ordering in the notemap.json? I tried inputting fingerings for low A and Ab as well as high G (maybe G# too), but it would not load. I did simply put the fingerings manually at the end of the list, would that be part of the problem or is the Haxophone somehow limited in the note range it accepts in the notemap?

I do not think order matters (have not checked), but all the fingerings must be unique:  i.e. many fingerings can exist for a note, but only one note can be associated with a given fingering.  In other words, the key (left) side of each line must be unique:

Not OK:
```
  "4826256": 63,
  "4826256": 64,
```
OK:
```
  "4826256": 63,
  "4826257": 63,
```

What error did you get when loading your notemap?  (`journalctl -u haxo`)

> Could I make a C-transposed notemap by simply offsetting the MIDI Notes in the notemap? Of course I would love to select it out of the dropdown menu on the web editor, but I might be able to script it for myself. With the coming changes to transposing, maybe this doesn't matter?

Yes offsetting the MIDI notes works.  In fact this is what Peter has in the works here https://github.com/petermoz/haxo-rs/blob/transpose-mode/notemap.json
Do not use the web editor.  It is a very hackish implementation I put together for my own use, and it's not very good.  A text editor is all you need.

> I'm working on some notemaps that are very large. Do you think these would affect the realtime performance of the Haxophone? One of the notemaps uses ca 25-30 fingerings for the same note. Another one as of now has just over 80 fingerings for the same note, though I don't know if it will go much above 90-100 fingerings per note.

No, I don't think so.  The notemap is converted into a binary search tree.  Looking up a fingering is really fast, regardless of the number of notes associated with a given fingering.

**Reply by Ekolide** [reply]
Wonderful! Thank you very much. I'll stop using the web editor, though it was very helpful for a bit, and certainly helpful as a tool to visualize the fingerings.

> What error did you get when loading your notemap? (journalctl -u haxo)

I threw away the notemap after a couple of attempts to not clutter my files with broken notemaps, and decided to return to it another day. I'll keep you updated on how it works out for me next time.

I did write down the specific keycodes as

```
Low A: "13444736":43,"13444752":43
Low Ab: "13448832";42,"13448848";42

High G: "1048937";77,"16745";77
High G#: "1065321";78
```

There might have been a conflict, I do not know, and I do not remember what these specifics fingerings meant in context, either.

> No, I don't think so. The notemap is converted into a binary search tree.

Perfect.

Thanks again!

Edit: I caught my error, I accidentally used semi-colons instead of colons for some of the keycodes above. Trying it on the Haxophone next.

Edit: Works fine on the Haxophone!

**Comment by Utopiah** [comment]
Here is mine, work in progress https://webmidi-abc-chromatone.glitch.me/haxo.html

**Comment by jcard0na** [comment]
> @jcard0na I've found that when deleting all fingerings for a note, then adding a new fingering destroys the UI with no apparent way to get it back.

I fixed this yesterday.  Let me know if you experience another similar crash.  Thanks,

**Reply by Ekolide** [reply]
Wonderful! Thanks so much.

I'm very interested to know how you "unwrap" the hardware codes. I've figured out the system of columns and rows on the Haxophone, and how it's generating it's respective output, entangling into giant numbers. What I'm less sure about is how all the additions can be untangled back, to separate into it's unique combination of buttons. Seems like some fundamental math or engineering principle at work?

Oh, and is the fingering-matrix.png and .svg in the repo flipped for C1:R2 and C1:R1? From your tool, the bis key outputs `16`, and the B key outputs `128`.

Thanks again for the work you do on the Haxophone, and answering all my inquires.

Edit: I figured it out, it's all Base-8

**Reply by jcard0na** [reply]
> I'm very interested to know how you "unwrap" the hardware codes. I've figured out the system of columns and rows on the Haxophone, and how it's generating it's respective output, entangling into giant numbers. What I'm less sure about is how all the additions can be untangled back, to separate into it's unique combination of buttons. Seems like some fundamental math or engineering principle at work?

Every key is represented by a single bit in a 24-bit number.  The key at row `r` and column `c` corresponds to bit `3*r + c`.  The binary number constructed from those bits is converted into decimal, and that is the "giant number" you are seeing.
The formula is:

```math
fingering = \sum_{c=0}^2 \sum_{r=0}^7 key\_pressed(r,c) \cdot 2^{3r + c} 
```
where `key_pressed(r,c)` is `1` if the key at row `r`, column `c` is pressed, or `0` otherwise.


So the low A fingering, which requires pressing L1 (`r=2`, `c=1`) and L2 (`r=3`, `c=1`) would be represented by number

```math
2^{2\cdot3+1} + 2^{3\cdot3+1} = 128 + 1024 = 1152 
```

You can confirm that here:  https://haxo-notemap.nn.r.appspot.com/
Toggle notemap visibility and you'll see that one of the `A` fingerings (midi 69) is "1152"

![image](https://github.com/cardonabits/haxo-rs/assets/676181/2132c22e-6d7e-42c0-a4c9-6695e1a4af58)

To convert back you can just convert the number into binary and inspect which bit positions are 1:

```
1152 = 0b10010000000
```
That's positions 7 and 10 (counting from the left), which correspond to:

 ```math 
r = floor (7 / 3) = 2, c = 7 \bmod 3 = 1
```
and
```math
r = floor (10 / 3) = 3, c = 7 \bmod 3 = 1
```

> Oh, and is the fingering-matrix.png and .svg in the repo flipped for C1:R2 and C1:R1? From your tool, the bis key outputs 16, and the B key outputs 128.

The bis key is flipped yes.  This is because the .svg file is based on standard sax fingering diagrams, where the bis key is shown between R1 and R2 (R1 here meaning "right first finger", not "first row"), shifted to the right.  See an example of a fingering diagram below:

![image](https://github.com/cardonabits/haxo-rs/assets/676181/ecfc51ab-abea-46f8-860b-4a1359a989cf)

On the haxophone, the bis key actually is **on the left** of R1 (this is the result of countless iterations to adapt the ergonomics of a saxophone to use computer keys 😅 ).  As the [.png image](https://haxo-notemap.nn.r.appspot.com/sax_fingering.png) represents a real haxophone layout, the bis key is on the left, the opposite side.

cc: @Utopiah

**Reply by jcard0na** [reply]
This diagram might help clarify how to map row numbers with actual keys on the haxophone:

![sax_fingering](https://github.com/cardonabits/haxo-rs/assets/676181/5a85bd81-94c9-4bc6-80fe-b050fae39025)

**Reply by Ekolide** [reply]
Hey Javier, thank you so much for the detailed answer. It all helps a ton. This seems like a comment that could go unedited into whatever documentation might be suitable.

Indeed I was getting confused about the fingering matrix, even though I was clear in my head that 1 is the bis key and 2 was supposed to be the B key. Thanks you for the helpful graphic.

**Reply by Ekolide** [reply]
Hiya. New bug: When removing fingers that are already in the currently selected fingering, the UI doesn't update. I can still remove fingers from the fingering and save it as a new one, but it won't update in real time. Hope this makes sense.

**Reply by jcard0na** [reply]
> When removing fingers that are already in the currently selected fingering, the UI doesn't update.

I just fixed it.  Thanks for reporting!
